cli_tools/gce_vm_image_import/importer/inflater.go

```diff
@@ -40,6 +43,119 @@ type inflater interface {
 	traceLogs() []string
 }
 
+// inflaterFacade implements an inflater using other concrete implementations.
+type inflaterFacade struct {
+	mainInflater   inflater
+	shadowInflater inflater
+}
+
+// signals to control the verification towards shadow inflater
+const (
+	sigMainInflaterDone   = "main done"
+	sigMainInflaterErr    = "main err"
+	sigShadowInflaterDone = "shadow done"
+	sigShadowInflaterErr  = "shadow err"
+)
+
+func (facade inflaterFacade) inflate(ctx context.Context) (persistentDisk, error) {
+	inflaterChan := make(chan string)
+
+	// Launch main inflater.
+	var pd persistentDisk
+	var err error
+	go func() {
+		pd, err = facade.mainInflater.inflate(ctx)
+		if err != nil {
+			inflaterChan <- sigMainInflaterErr
+		} else {
+			inflaterChan <- sigMainInflaterDone
+		}
+	}()
+
+	if facade.shadowInflater == nil {
+		pd.matchResult = "NA"
+		// Wait for main inflater to finish
+		<-inflaterChan
+		return pd, err
+	}
+
+	// Launch shadow inflater.
+	var shadowPd persistentDisk
+	var shadowErr error
+	go func() {
+		shadowPd, shadowErr = facade.shadowInflater.inflate(ctx)
+		if shadowErr != nil {
+			inflaterChan <- sigShadowInflaterErr
+		} else {
+			inflaterChan <- sigShadowInflaterDone
+		}
+	}()
```

Comment:
As I'm reading this, I'm curious about what happens if one or both inflaters keep going forever. My guess is that the children inflaters are responsible for managing their own lifecycle?

Reply:
This is a interesting topic.
For daisy inflater, it has a daisy timeout.
For API inflater, it relies on GCP operation (the long-running API handle) mechanism. We assume that waiting on 'operation' is safe and won't run forever.

We can also add an additional client-side timeout mechanism for the api inflater, however I chose not to do that, since the whole daisy workflow's timeout is based on the same assumption. For exmaple, step_create_instance. Even when daisy emit a timeout signal to the step_create_instance, it won't terminate at once. Instead, it wait for the API to finish. So, giving the API inflater a additional timeout won't obviously improve the whole import's safety. The whole import still mostly rely on the GCP API's mechanism.

In other words, one single API call is somehow 'atomic' for daisy.